addons/hr_payroll/models/hr_payslip.py

```diff
@@ -0,0 +1,581 @@
+    payslip_run_id = fields.Many2one('hr.payslip.run', 'Payslip Batches', readonly=True, states={'draft': [('readonly', False)]}, copy=False)
+    payslip_count = fields.Integer(compute='_compute_payslip_count', string='Payslip Computation Details')
+
+    @api.one
+    @api.constrains('date_from', 'date_to')
+    def check_dates(self):
+        if not self._check_dates():
+            raise ValidationError(
+                _("Payslip 'Date From create must be before 'Date To'"))
+
+    @api.multi
+    def _compute_payslip_count(self):
+        payslip_data = self.env['hr.payslip.line'].read_group([('slip_id', 'in', self.ids)], ['slip_id'], ['slip_id'])
+        mapped_data = dict([(m['slip_id'][0], m['slip_id_count']) for m in payslip_data])
+        for payslip in self:
+            payslip.payslip_count = mapped_data.get(payslip.id, 0)
+
+    @api.multi
+    def _compute_details_by_salary_rule_category(self):
+        result = {}
+        if not self.ids:
+            return result
+        for id in self.ids:
+            result.setdefault(id, [])
+        self.env.cr.execute('''SELECT pl.slip_id, pl.id FROM hr_payslip_line AS pl \
+                    LEFT JOIN hr_salary_rule_category AS sh on (pl.category_id = sh.id) \
+                    WHERE pl.slip_id in %s \
+                    GROUP BY pl.slip_id, pl.sequence, pl.id ORDER BY pl.sequence''',(tuple(self.ids),))
+        res = self.env.cr.fetchall()
+        for r in res:
+            result[r[0]].append(r[1])
+        return result
+
+    def _check_dates(self):
+        for payslip in self.filtered(lambda x: x.date_from > x.date_to):
+            return False
+        return True
+
+    @api.multi
+    def cancel_sheet(self):
+        return self.write({'state': 'cancel'})
+
+    @api.multi
+    def process_sheet(self):
+        return self.write({'paid': True, 'state': 'done'})
+
+    @api.multi
+    def hr_verify_sheet(self):
+        self.compute_sheet()
+        return self.write({'state': 'verify'})
+
+    @api.multi
+    def refund_sheet(self):
+        for payslip in self:
+            id_copy = payslip.copy({'credit_note': True, 'name': _('Refund: %s') % payslip.name})
+            payslip.signal_workflow('hr_verify_sheet')
+            payslip.signal_workflow('process_sheet')
+
+        form_res = self.env.ref('hr_payroll.view_hr_payslip_form').id or False
+        tree_res = self.env.ref('hr_payroll.view_hr_payslip_tree').id or False
+        return {
+            'name':_("Refund Payslip"),
+            'view_mode': 'tree, form',
+            'view_id': False,
+            'view_type': 'form',
+            'res_model': 'hr.payslip',
+            'type': 'ir.actions.act_window',
+            'target': 'current',
+            'domain': "[('id', 'in', %s)]" % [id_copy.id],
+            'views': [(tree_res, 'tree'), (form_res, 'form')],
+            'context': {}
+        }
+
+    def check_done(self):
```

Comment:
useless no need of this function as this function is called in workflow transition condition.. and bydefault 
workflow transition condition field value is True.....